### PR TITLE
Use OccupiesSpace to avoid Mobile look-ups

### PR DESCRIPTION
**Pull request by reaperrr**
in `PathFinder.FindUnitPath` and `.FindUnitPathToRange`.

RA shellmap on bleed:
```
lst 418 Find*Path Mobile look-up time: 00:00:00.0000048
harv 519 Find*Path Mobile look-up time: 00:00:00.0000004
harv 520 Find*Path Mobile look-up time: 00:00:00.0000002
3tnk 445 Find*Path Mobile look-up time: 00:00:00.0000011
3tnk 454 Find*Path Mobile look-up time: 00:00:00.0000002
3tnk 463 Find*Path Mobile look-up time: 00:00:00.0000001
3tnk 472 Find*Path Mobile look-up time: 00:00:00.0000002
3tnk 481 Find*Path Mobile look-up time: 00:00:00.0000001
3tnk 490 Find*Path Mobile look-up time: 00:00:00.0000002
e1 499 Find*Path Mobile look-up time: 00:00:00.0000001
e2 500 Find*Path Mobile look-up time: 00:00:00.0000004
e3 501 Find*Path Mobile look-up time: 00:00:00.0000004
e4 502 Find*Path Mobile look-up time: 00:00:00.0000006
ftrk 446 Find*Path Mobile look-up time: 00:00:00.0000003
ftrk 455 Find*Path Mobile look-up time: 00:00:00.0000001
ftrk 464 Find*Path Mobile look-up time: 00:00:00.0000002
ftrk 473 Find*Path Mobile look-up time: 00:00:00.0000002
ftrk 482 Find*Path Mobile look-up time: 00:00:00.0000002
ftrk 491 Find*Path Mobile look-up time: 00:00:00.0000001
e1 503 Find*Path Mobile look-up time: 00:00:00.0000004
e1 521 Find*Path Mobile look-up time: 00:00:00.0000010
e1 521 Find*Path Mobile look-up time: 00:00:00.0000003
e1 521 Find*Path Mobile look-up time: 00:00:00.0000001
e1 521 Find*Path Mobile look-up time: 00:00:00.0000008
e1 521 Find*Path Mobile look-up time: 00:00:00.0000004
e2 504 Find*Path Mobile look-up time: 00:00:00.0000005
e3 505 Find*Path Mobile look-up time: 00:00:00.0000004
ttnk 447 Find*Path Mobile look-up time: 00:00:00.0000003
ttnk 456 Find*Path Mobile look-up time: 00:00:00.0000002
ttnk 465 Find*Path Mobile look-up time: 00:00:00.0000002
ttnk 474 Find*Path Mobile look-up time: 00:00:00.0000002
ttnk 483 Find*Path Mobile look-up time: 00:00:00.0000001
ttnk 492 Find*Path Mobile look-up time: 00:00:00.0000002
e4 506 Find*Path Mobile look-up time: 00:00:00.0000003
e1 507 Find*Path Mobile look-up time: 00:00:00.0000003
```

RA shellmap on this PR:
```
lst 418 Find*Path Mobile look-up time: 00:00:00.0000019
harv 519 Find*Path Mobile look-up time: 00:00:00.0000003
harv 520 Find*Path Mobile look-up time: 00:00:00.0000001
3tnk 445 Find*Path Mobile look-up time: 00:00:00.0000003
3tnk 454 Find*Path Mobile look-up time: 00:00:00.0000001
3tnk 463 Find*Path Mobile look-up time: 00:00:00
3tnk 472 Find*Path Mobile look-up time: 00:00:00
3tnk 481 Find*Path Mobile look-up time: 00:00:00.0000001
3tnk 490 Find*Path Mobile look-up time: 00:00:00
e1 499 Find*Path Mobile look-up time: 00:00:00.0000001
e2 500 Find*Path Mobile look-up time: 00:00:00.0000003
e3 501 Find*Path Mobile look-up time: 00:00:00.0000003
e4 502 Find*Path Mobile look-up time: 00:00:00.0000002
ftrk 446 Find*Path Mobile look-up time: 00:00:00.0000003
ftrk 455 Find*Path Mobile look-up time: 00:00:00
ftrk 464 Find*Path Mobile look-up time: 00:00:00.0000001
ftrk 473 Find*Path Mobile look-up time: 00:00:00.0000001
ftrk 482 Find*Path Mobile look-up time: 00:00:00
ftrk 491 Find*Path Mobile look-up time: 00:00:00.0000001
e1 503 Find*Path Mobile look-up time: 00:00:00.0000002
e1 521 Find*Path Mobile look-up time: 00:00:00.0000002
e1 521 Find*Path Mobile look-up time: 00:00:00
e1 521 Find*Path Mobile look-up time: 00:00:00
e1 521 Find*Path Mobile look-up time: 00:00:00.0000001
e1 521 Find*Path Mobile look-up time: 00:00:00.0000006
e2 504 Find*Path Mobile look-up time: 00:00:00.0000002
e3 505 Find*Path Mobile look-up time: 00:00:00.0000002
ttnk 447 Find*Path Mobile look-up time: 00:00:00.0000002
ttnk 456 Find*Path Mobile look-up time: 00:00:00.0000001
ttnk 465 Find*Path Mobile look-up time: 00:00:00.0000001
ttnk 474 Find*Path Mobile look-up time: 00:00:00
ttnk 483 Find*Path Mobile look-up time: 00:00:00
ttnk 492 Find*Path Mobile look-up time: 00:00:00.0000001
e4 506 Find*Path Mobile look-up time: 00:00:00.0000002
e1 507 Find*Path Mobile look-up time: 00:00:00.0000002
```

Not much, but still measurable and with minimal code change.

### CODE REVIEW

**Review by Mailaender — DISMISSED** [review]
I'm sold.

File: OpenRA.Mods.Common/Traits/World/PathFinder.cs [hunk 1]
```diff
@@ -64,7 +64,8 @@ public PathFinder(World world)
 
 		public List<CPos> FindUnitPath(CPos source, CPos target, Actor self, Actor ignoreActor, BlockedByActor check)
 		{
-			var mobile = self.Trait<Mobile>();
+			// PERF: Because we can be sure that OccupiesSpace is Mobile here, we can save some performance by avoiding querying for the trait.
+			var mobile = self.OccupiesSpace as Mobile;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/PathFinder.cs]: It is more efficient to cast directly, like `var locomotor = ((Mobile)self.OccupliesSpace).Locomotor`.

**Comment by reaperrr** [comment]
Updated.

**Review by atlimit8 — APPROVED** [review]

**Comment by atlimit8** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/2ce44cce3476fe43a8182442bd1d9f5bdb47b4fb)

**Pull request merged**